### PR TITLE
Fix delta comparison for auto failover, multi AZ, primary cluster ID

### PR DESCRIPTION
**Pull request by echen-98**
Nice.

Description of changes:
Properly retrieve the latest state for these 3 fields and compare them to the desired state. For each of these 3 spec fields, obtaining the corresponding latest state requires processing of a different `Status` field due to the structure of the `DescribeReplicationGroups` API response

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: test/e2e/tests/test_replicationgroup.py [hunk 1]
```diff
@@ -607,8 +606,8 @@ def test_rg_fault_tolerance(self, rg_fault_tolerance):
                 raise AssertionError(f"Unknown node {node['cacheClusterID']}")
 
         # assert AF and multi AZ
-        assert resource['status']['automaticFailover'] == "disabled"
-        assert resource['status']['multiAZ'] == "disabled"
+        assert resource['status']['automaticFailover'] == "enabled"
```
Comment (echen-98) [inline test/e2e/tests/test_replicationgroup.py:609]: these were just wrong assertions; if you check the file the last patch enables both of these fields

File: pkg/resource/replication_group/custom_update_api.go [hunk 2]
```diff
@@ -71,27 +71,9 @@ func (rm *resourceManager) CustomModifyReplicationGroup(
 	}
 
 	// Order of operations when diffs map to multiple updates APIs:
-	// 1. When automaticFailoverEnabled differs:
-	//		if automaticFailoverEnabled == false; do nothing in this custom logic, let the modify execute first.
-	// 		else if automaticFailoverEnabled == true then following logic should execute first.
-	// 2. When multiAZ differs
-	// 		if multiAZ = true  then below is fine.
-	// 		else if multiAZ = false ; do nothing in custom logic, let the modify execute.
```
[the remark this answers is not in the document]
Comment (echen-98) [reply pkg/resource/replication_group/custom_update_api.go:79]: > When desired value is false then invoke modifyReplicationGroup (default implementation in sdk.go) first

`modifyReplicationGroup` only leads to the default/generated implementation if it returns `nil, nil`, which happens when there is no difference in security group IDs or engine version. That is to say, if we want to skip straight to the generated code, we should immediately `return nil, nil` instead of call `rm.modifyReplicationGroup`. 

> When desired value is true then let the rest of the custom modify logic execute first for AF/multiAZ to be applied later once all custom logic has executed.

This second case is equivalent to removing the below block, so the uncertainty remains with the first case above. Which of the below fields require multi AZ and auto failover to be disabled first (replica count is one, but I'm not sure if there's another)? If order matters then we should have had a test which failed upon removing this block, but we can add one now if it doesn't exist.

[the remark this answers is not in the document]
Comment (echen-98) [reply pkg/resource/replication_group/custom_update_api.go]: Went ahead and reverted the changes to this particular file. Since we have limited time to iterate on this (as discussed offline), we can hold these thoughts for later. Same with non-blocking comments below, let's merge if this test run passes.

File: pkg/resource/replication_group/delta_util.go [hunk 3]
```diff
@@ -100,3 +113,76 @@ func unmarshalLastRequestedLDCs(desired *resource) []*svcapitypes.LogDeliveryCon
...
+		return true
+	}
+
+	// true maps to "enabled"; false maps to "disabled"
+	// this accounts for values such as "enabling" and "disabling"
+	if *desired.ko.Spec.MultiAZEnabled {
+		return *latest.ko.Status.MultiAZ != string(svcapitypes.MultiAZStatus_enabled)
+	} else {
+		return *latest.ko.Status.MultiAZ != string(svcapitypes.MultiAZStatus_disabled)
+	}
```
Comment (jaypipes) [inline pkg/resource/replication_group/delta_util.go:136]: Two things:

1) There should not be a `Status.MultiAZ` field. There should only be a single field: `Spec.MultiAZEnabled`

2) This kind of custom code should go in your [`updateSpecFields`](https://github.com/aws-controllers-k8s/elasticache-controller/blob/d3a3ba8e3a8da76b4b16a37b09e5675111e7f3c9/pkg/resource/replication_group/post_set_output.go#L28) function in `custom_set_output.go` and look something like this:

```go
	// MultiAZEnabled is a boolean in the Spec field but a string enum in the
	// output shape field. Here, we reconcile that difference...
	if (respRG.MultiAZ == nil) {
		// Or set this to a default value if nil isn't default?
		resource.ko.Spec.MultiAZEnabled = nil
	} else {
		resource.ko.Spec.MultiAZEnabled = (*respRG.MultiAZ == string(svcapitypes.MultiAZStatus_enabled))
	}
```

If you do the above, you don't have to worry about delta comparisons at all.

**Review by jaypipes — CHANGES_REQUESTED** [review]

File: pkg/resource/replication_group/delta_util.go (hunk 3 above)
Comment (echen-98) [reply pkg/resource/replication_group/delta_util.go:136]: > There should not be a Status.MultiAZ field. There should only be a single field: Spec.MultiAZEnabled

Agreed in principle, but the existence of a separate `MultiAZ` field is an ElastiCache API quirk rather than a conscious decision on our part.

> This kind of custom code should go in your updateSpecFields function in custom_set_output.go and look something like this:

I'm glad you brought this up because it's helping me externalize why we're moving away from `updateSpecFields` in many cases. 

Initially, I wanted to use `updateSpecFields` to update every single spec field during `ReadOne`, so that the generated delta code could handle everything and there would be no need for manual comparison between `desired.Spec` and `latest.Status`. 

However, `Spec` fields often don't allow us to store all the information we're getting from the API response. For example, `Spec.AutomaticFailoverEnabled` is a bool, whose associated field `Status.AutomaticFailover` is a string/enum which could be any of "enabled", "disabled", "enabling", or "disabling". We would have to map each of these four states back to a bool (square peg, round hole sort of thing), when it would make more sense to just make a decision based on the full version of the information (`Status`) since we have it anyway.

Granted, the `MultiAZ` case is a little different since at the moment, the only two possible values are "enabled" and "disabled". However, it's entirely possible that in the future ElastiCache decides to have this field update asynchronously or make any other change which could introduce additional states/values for `Status.MultiAZ`. So the suggested approach could break given these changes, and even if no service-side changes occur, I would say it's still non-ideal to force a string/enum into a bool when we can just make a decision based on the full, unadulterated information from the API response.

As far as the role of `updateSpecFields`, I think it absolutely should exist, but the function name and intended purpose should change. The majority of this function is already [calling the DescribeCacheClusters API](https://github.com/aws-controllers-k8s/elasticache-controller/blob/main/pkg/resource/replication_group/post_set_output.go#L41-L47) and populating the associated fields in the replication group object, since the latest state for some `Spec` fields cannot be retrieved directly from the `DescribeReplicationGroups API`. I would say that this should be the primary purpose of this function in the future.

So to sum up, the function currently named `updateSpecFields` is an extension of `SdkFind`/`ReadOne` in that we are still retrieving the latest state of the replication group (but via a different API). There should be limited mapping back to `Spec` fields to avoid losing information; we only do it here because this state will be lost if we don't save it back to the `latest` object. On the other hand, `modifyDelta` should look at the most complete representation of the latest state of each `Spec` field available to it (which is often in the `Status`) and make a decision based on that. Therefore I think we should keep things as-is.

File: pkg/resource/replication_group/delta_util.go [hunk 4]
```diff
@@ -54,6 +54,19 @@ func filterDelta(
 		delta.Add("Spec.LogDeliveryConfigurations", desired.ko.Spec.LogDeliveryConfigurations,
 			unmarshalLastRequestedLDCs(desired))
 	}
+
+	if multiAZRequiresUpdate(desired, latest) {
+		delta.Add("Spec.MultiAZEnabled", desired.ko.Spec.MultiAZEnabled, latest.ko.Status.MultiAZ)
```
[the remark this answers is not in the document]
Comment (echen-98) [reply pkg/resource/replication_group/delta_util.go:59]: They are different types; see response below for more detail
